src/classes/RD2_ApiService.cls

```diff
@@ -0,0 +1,178 @@
+/*
+    Copyright (c) 2020 Salesforce.org
+    All rights reserved.
+
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of Salesforce.org nor the names of
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2020
+* @group Recurring Donations
+* @description All services specific to Enhanced Recurring Donation Api's
+*
+*/
+public inherited sharing class RD2_ApiService {
+
+    /**
+     * @description Retrieve a map of RecurringDonationSchedule__c records by Rd.id using an inner wrapper
+     * class representation of the RdSchedule object because it's protected and customers will not be able to
+     * deserialize it.
+     * @param rdIds Recurring Donation Id's to retrieve schedules for
+     * @return JSON Serialized version of Map<Id, List<RD2_ApiService.Schedule>> schedulesByRdId
+     */
+    public String querySchedules(Set<Id> rdIds) {
+        Map<Id, List<RD2_ApiService.Schedule>> schedulesByRdId = new Map<Id, List<RD2_ApiService.Schedule>>();
+
+        if (!RD2_EnablementService.isRecurringDonations2Enabled) {
+            return JSON.serialize(schedulesByRdId);
+        }
+
+        RD2_ScheduleService.PauseScheduleHandler pauseHandlerService = new RD2_ScheduleService.PauseScheduleHandler();
+
+        String soql = getRecurringDonationQuery();
+        for (npe03__Recurring_Donation__c rd : Database.query(soql)) {
+
+            Boolean isRdCurrentlyPaused = pauseHandlerService.hasActivePause(
```

Comment:
When a pause is valid (Start Date <= End Date) and in future (referenceDate <= End Date), then the pause schedule is considered as an active pause. However, it does not necessarily mean it is the current pause. Similarly is true for a regular active schedule: we can have up to two active future schedules but only one is current.  In summary, neither `isActive` nor `isActivePause`  conveys if the schedule is current, ie neither method checks if "Start Date <= referenceDate <= EndDate". 

So we need another method,  or order of commands to find the active pause schedule, and if one exists, check if the current schedule is paused. Otherwise, if no pause exists, then no additional check is to be done.

Reply:
Good catch. I misread the meaning and logic in that method. The method description is `Indicates if the Recurring Donation is an active pause schedule` which I read has the RD is currently paused. I'll update the description to clarify.

I do agree that another method is needed to see if the pause is currently active.

Reply:
I'll add that the `isWithinDateRange` method name is very misleading. If I provide a date to that method I would expect it to tell me if that date is within the range of the start and end dates on the record. That is definitely not what it does though. It actually returns true if the record is either current or in the future based on the provided date. 

What would you think of me renaming the method to `isCurrentOrFuture()`? Note that this method is duplicated in another place in the class so I would rename both.
@npsp-reedestockton

Reply:
I have not checked the code for existing method, wondering if we can leave it as is, or it is required to be renamed. With regards to renaming `isWithingDateRange()`, it is perfectly fine and valid. How about we use `isScheduled()` instead ("scheduled" is one of the synonyms of "future" :)))

Reply:
It's not *required* to rename, but I definitely feel that the method name does not reflect the behavior. I like `isScheduled` as a replacement for `isWithinDateRange`.